File: cli/src/commands/list.ts
import fs from 'node:fs';
import path from 'node:path';
import { findProjectRoot, loadYaml } from '../utils.js';

export function runList(): void {
  const root = findProjectRoot();

  // ── Agents ──────────────────────────────────────────────────────────────
  const agentsDir = path.join(root, 'agents');
  const agents: Array<{
    id: string;
    type: string;
    model: string | null;
    hasPrompt: boolean;
  }> = [];

  if (fs.existsSync(agentsDir)) {
    for (const entry of fs.readdirSync(agentsDir, { withFileTypes: true })) {
      if (!entry.isDirectory()) continue;
      const yamlPath = path.join(agentsDir, entry.name, 'agent.yaml');
      if (!fs.existsSync(yamlPath)) continue;

      const data = loadYaml(yamlPath);
      agents.push({
        id: entry.name,
        type: data.type ?? 'unknown',
        model: data.model ?? null,
        hasPrompt: fs.existsSync(path.join(agentsDir, entry.name, 'prompt.md')),
      });
    }
  }

  // ── Workflows ───────────────────────────────────────────────────────────
  const workflowsDir = path.join(root, 'workflows');
  const workflows: Array<{
    name: string;
    version: string;
    stepCount: number;
    stepIds: string[];
  }> = [];

  if (fs.existsSync(workflowsDir)) {
    for (const f of fs.readdirSync(workflowsDir)) {
      if (!f.endsWith('.yaml') && !f.endsWith('.yml')) continue;
      const data = loadYaml(path.join(workflowsDir, f));
      const name = f.replace(/\.ya?ml$/, '');
      const steps: any[] = Array.isArray(data.steps) ? data.steps : [];
      workflows.push({
        name,
        version: data.version ?? '-',
        stepCount: steps.length,
        stepIds: steps.map((s: any) => s.id ?? '?'),
      });
    }
  }

  // ── Print ───────────────────────────────────────────────────────────────
  if (agents.length === 0 && workflows.length === 0) {
    console.log('No agents or workflows found. Run `agentic init` to get started.');
    return;
  }

  if (agents.length > 0) {
    console.log('Agents');
    console.log('------');
    for (const a of agents) {
      const modelStr = a.model ? ` model=${a.model}` : '';
      const promptStr = a.hasPrompt ? ' prompt.md=yes' : '';
      console.log(`  ${a.id}  type=${a.type}${modelStr}${promptStr}`);
    }
    console.log();
  }

  if (workflows.length > 0) {
    console.log('Workflows');
    console.log('---------');
    for (const w of workflows) {
      console.log(
        `  ${w.name}  version=${w.version}  steps=${w.stepCount} [${w.stepIds.join(', ')}]`,
      );
    }
    console.log();
  }
}

File: cli/src/commands/list.rs
use std::fs;

use crate::utils::{find_project_root, load_yaml};

pub fn run() {
    let root = find_project_root();

    // ── Agents ──────────────────────────────────────────────────────────────
    let agents_dir = root.join("agents");
    let mut agents: Vec<(String, String, Option<String>, bool)> = Vec::new();

    if agents_dir.exists() {
        if let Ok(entries) = fs::read_dir(&agents_dir) {
            for entry in entries.flatten() {
                if !entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                    continue;
                }
                let name = entry.file_name().to_string_lossy().to_string();
                let yaml_path = entry.path().join("agent.yaml");
                if !yaml_path.exists() {
                    continue;
                }
                let data = load_yaml(&yaml_path);
                let agent_type = data
                    .get("type")
                    .and_then(|v| v.as_str())
                    .unwrap_or("unknown")
                    .to_string();
                let model = data
                    .get("model")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string());
                let has_prompt = entry.path().join("prompt.md").exists();
                agents.push((name, agent_type, model, has_prompt));
            }
        }
    }
    agents.sort_by(|a, b| a.0.cmp(&b.0));

    // ── Workflows ───────────────────────────────────────────────────────────
    let workflows_dir = root.join("workflows");
    let mut workflows: Vec<(String, String, usize, Vec<String>)> = Vec::new();

    if workflows_dir.exists() {
        if let Ok(entries) = fs::read_dir(&workflows_dir) {
            for entry in entries.flatten() {
                let fname = entry.file_name().to_string_lossy().to_string();
                if !fname.ends_with(".yaml") && !fname.ends_with(".yml") {
                    continue;
                }
                let data = load_yaml(&entry.path());
                let name = fname
                    .trim_end_matches(".yaml")
                    .trim_end_matches(".yml")
                    .to_string();
                let version = data
                    .get("version")
                    .and_then(|v| v.as_str())
                    .unwrap_or("-")
                    .to_string();
                let steps = data
                    .get("steps")
                    .and_then(|v| v.as_sequence())
                    .map(|seq| {
                        seq.iter()
                            .map(|s| {
                                s.get("id")
                                    .and_then(|v| v.as_str())
                                    .unwrap_or("?")
                                    .to_string()
                            })
                            .collect::<Vec<_>>()
                    })
                    .unwrap_or_default();
                let step_count = steps.len();
                workflows.push((name, version, step_count, steps));
            }
        }
    }
    workflows.sort_by(|a, b| a.0.cmp(&b.0));

    // ── Print ───────────────────────────────────────────────────────────────
    if agents.is_empty() && workflows.is_empty() {
        println!("No agents or workflows found. Run `agentic init` to get started.");
        return;
    }

    if !agents.is_empty() {
        println!("Agents");
        println!("------");
        for (id, agent_type, model, has_prompt) in &agents {
            let model_str = model
                .as_ref()
                .map(|m| format!(" model={}", m))
                .unwrap_or_default();
            let prompt_str = if *has_prompt { " prompt.md=yes" } else { "" };
            println!("  {}  type={}{}{}", id, agent_type, model_str, prompt_str);
        }
        println!();
    }

    if !workflows.is_empty() {
        println!("Workflows");
        println!("---------");
        for (name, version, step_count, step_ids) in &workflows {
            println!(
                "  {}  version={}  steps={} [{}]",
                name,
                version,
                step_count,
                step_ids.join(", ")
            );
        }
        println!();
    }
}
